src/compact_live.rs

```diff
@@ -0,0 +1,449 @@
+///
+/// The following operations have been applied:
+/// - a new random ephemeral key `K_w'` is drawn;
+/// - the new last chain UID is stored in the Entry Table entry;
+/// - chain UIDs are generated anew (random-like);
+/// - chain values are encrypted under the new ephemeral key;
+/// - `B1` is removed since its last occurrence in the chain is a deletion;
+/// - `B2` is kept since its last occurrence is an addition and is simplified
+///   (only one occurrence in the chain);
+/// - `B3` and `B4` are kept since they have not been deleted, and simplified.
+///
+/// *NOTE*: the order of the locations in the compacted chain is not guaranteed.
+pub trait FindexLiveCompact<
+    const UID_LENGTH: usize,
+    const BLOCK_LENGTH: usize,
+    const CHAIN_TABLE_WIDTH: usize,
+    const MASTER_KEY_LENGTH: usize,
+    const KWI_LENGTH: usize,
+    const KMAC_KEY_LENGTH: usize,
+    const DEM_KEY_LENGTH: usize,
+    KmacKey: SymKey<KMAC_KEY_LENGTH>,
+    DemScheme: Dem<DEM_KEY_LENGTH>,
+    CustomError: std::error::Error + CallbackError,
+>:
+    FindexCallbacks<CustomError, UID_LENGTH>
+    // to get `batch_fetch_chains()`
+    + FindexCompact<
+        UID_LENGTH,
+        BLOCK_LENGTH,
+        CHAIN_TABLE_WIDTH,
+        MASTER_KEY_LENGTH,
+        KWI_LENGTH,
+        KMAC_KEY_LENGTH,
+        DEM_KEY_LENGTH,
+        KmacKey,
+        DemScheme,
+        CustomError,
+    >
+{
+    /// Ratio of the noise over the real data.
+    const NOISE_RATIO: f64;
+
+    /// Number of noisy Entry Table UIDs to compact in a row.
+    const BATCH_SIZE: usize;
+
+    /// Selects a set of Entry Table UIDs to reindex (target set) and a set of Entry Table UIDs
+    /// used as noise (noise set).
+    ///
+    /// These sets are mutually exclusive and the ratio of the cardinal of the noise set over the
+    /// target set is given by `Self::NOISE_RATIO`.
+    ///
+    /// Returns the noisy taget set (target set + noise set) and the noise set.
+    /// A HashSet is used for the noise set because lookup should be fast.
+    ///
+    /// # Parameters
+    ///
+    /// - `rng`                             : secure random number generator
+    /// - `num_reindexing_before_full_set`  : number of compact operations need to compact all the
+    /// base
+    async fn select_noisy_uids(
+        &self,
+        rng: &mut impl CryptoRngCore,
+        num_reindexing_before_full_set: u32
+    ) -> Result<(Vec<Uid<UID_LENGTH>>, HashSet<Uid<UID_LENGTH>>), Error<CustomError>> {
+        if Self::NOISE_RATIO > 1f64 {
+            return Err(Error::CryptoError(format!(
+                "noise ration cannot be greater than 1 ({} given)",
+                Self::NOISE_RATIO
+            )));
+        }
+
+        let uids = self.fetch_all_entry_table_uids().await?;
+
+        // TODO (TBZ): find which probability is associated to this formula.
+        let entry_table_length = uids.len() as f64;
+        let n_compact = ((entry_table_length * (entry_table_length.log2() + 0.58))
```

Comment:
Magic number - related to the TODO ?

Reply:
This is the formula used to compute the size of the samples to draw in order to have drawn all the set after `num_reindexing_before_full_set` samples. This formula contains "magic numbers" but you are right, I will link an explanation (this was partly was I meant when writing my todo).

Reply:
I was thinking more like defining a constant with an explicit name to avoid useless cognitive complexity for the code.

Reply:
Done in ee25dc7